File: C++/src/algorithms/solveDiophantine.cpp
#include <iostream>
#include <stdexcept>
#include <vector>
#include <limits>
#include <tuple>
#include <cmath>
#include <gmpxx.h>

#include "mathUtils.hpp"
#include "menuUtils.hpp"
#include "algorithms.hpp"

using namespace std;

void solveDiophantine(mpz_class a, mpz_class b, mpz_class c, mpz_class& x, mpz_class& y, vector<mpz_class>& chainFractions) {
    /* Находит x, y для уравнения вида ax + by = d */
    mpz_class gcd = findGcd(findGcd(a, b), c);
    mpz_class finalCoeff = c / gcd;

    a /= gcd;
    b /= gcd;
    c /= gcd;

    mpz_class numerator = a;
    mpz_class denominator = b;

    int termCount = 0;
    while (denominator != 1) {
        chainFractions.push_back(numerator / denominator);

        mpz_class remains = numerator % denominator;
        
        numerator = denominator;
        denominator = remains;
        termCount++;
    }

    mpz_class finalNumerator = 1;
    mpz_class finalDenominator = chainFractions[termCount - 1];
    for (int i = chainFractions.size() - 2; i >= 0; i--) {
        finalNumerator = chainFractions[i] * finalDenominator + finalNumerator;
        mpz_class temp = finalDenominator;
        finalDenominator = finalNumerator;
        finalNumerator = temp;
    }

    x = pow(-1, termCount - 1) * finalNumerator * finalCoeff;
    y = pow(-1, termCount) * finalDenominator * finalCoeff;
}

void solveDioph() {
    mpz_class a = 275, b = 145, d = 10;
    mpz_class x = 0, y = 0;

    vector<mpz_class> chainFractions;

    string inputError = "Невозможно решить при заданных значениях.";
    try {
        solveDiophantine(a, b, d, x, y, chainFractions);
        cout << "Найденный X: " << x << endl;
        cout << "Найденный Y: " << y << endl;

        cout << "Цепная дробь: [";
        for (size_t i = 0; i < chainFractions.size(); i++) {
            cout << chainFractions[i];
            if (i == 0) {
                cout << "; ";
            } else if (i != chainFractions.size() - 1) {
                cout << ", ";
            }
        }
        cout << "]" << endl;
        
        cout << "Решение уравнения: " << a << "*x + " << b << "*y = " << d << ":\n";
        cout << "x = " << x << "\ny = " << y << endl << endl;
    } catch(const exception& e) {
        userInputError(e, inputError);
    }
}

File: GoLang/algorithms/solveDiophantine.go
package algorithms

import (
	"fmt"
	"math"
	"practice2/utils"
)

func SolveDiophantine(a, b, d int64) (int64, int64, []int64) {
	/* Находит x, y для уравнения ax + by = d и цепную дробь a / b */
	gcd := utils.FindGcd(utils.FindGcd(a, b), d)
	finalCoeff := d / gcd

	a /= gcd
	b /= gcd
	d /= gcd

	numerator := a
	denominator := b
	var cf []int64

	termCount := 0
	for denominator != 1 {
		cf = append(cf, numerator / denominator)

		remains := numerator % denominator

		numerator = denominator
		denominator = remains
		termCount++
	}

	finalNumerator := int64(1)
	finalDenominator := cf[termCount - 1]
	for _, val := range cf {
		finalNumerator = val * finalDenominator + finalNumerator
		temp := finalDenominator
		finalDenominator = finalNumerator
		finalNumerator = temp
	}

	x := int64(math.Pow(-1, float64(termCount - 1))) * finalNumerator * finalCoeff
	y := int64(math.Pow(-1, float64(termCount))) * finalDenominator * finalCoeff

	return x, y, cf
}

func SolveDioph() {
	var a int64 = 275
	var b int64 = 145
	var d int64 = 10

	x, y, chainFractions := SolveDiophantine(a, b, d)
	
	fmt.Printf("Найденный X: %d \n", x)
	fmt.Printf("Найденный Y: %d \n", y)

	fmt.Print("Цепная дробь: [")
	for i, val := range chainFractions {
		fmt.Print(val)
		if i == 0 {
			fmt.Print("; ")
		} else if i != len(chainFractions) - 1 {
			fmt.Print(", ")
		}
	}
	fmt.Println("]")

	fmt.Printf("Решение уравнения: %d*x + %d*y = %d:\n", a, b, d)
	fmt.Printf("x = %d \ny = %d \n\n", x, y)
}
